gtksourceview: init module

brckd opened the pull request:
Adds the `gtksourceview` target. This is used to configure GNOME Text Editor[^1], among others. The `gnome-text-editor` target is made mostly obsolete by this. It only works on Home Manager though.

Moves the theme template from the `gedit` module, because GTKSourceView is the more generic target.

*I haven't tested this yet!*

Closes: https://github.com/nix-community/stylix/issues/954

[^1]: https://github.com/danth/stylix/issues/954#issuecomment-2701605349

awwpotato reviewed (approved):
LGTM

brckd commented:
@danth I think you might be interested in testing this!

danth commented:
[This documentation page](https://wiki.gnome.org/Projects(2f)GtkSourceView(2f)StyleSchemes.html) suggests to me that a package containing `share/gnome/gtksourceview-{2.0,3.0}/styles/stylix.xml` might appear on the search path, which would get this working at the NixOS level too. If not, it contains a Python script to find out exactly what the search path is.

danth commented:
This works in my config!

There are some differences between the highlighting in GNOME Text Editor compared to VSCode, which could be caused by differences in our templates or differences in the editors themselves, however that's not relevant to this PR. The good thing is that it's now using the correct user level theme.

Once NixOS is supported, we should remove the `gnome-text-editor` module completely and just use `lib.mkRenamedOptionModule` to handle its enable option.

Do you want to work on NixOS now, or merge this and leave it for later?

brckd commented:
> There are some differences between the highlighting in GNOME Text Editor compared to VSCode, which could be caused by differences in our templates or differences in the editors themselves, however that's not relevant to this PR.

We could adapt both themes to the [base16 style guide](https://github.com/chriskempson/base16/blob/main/styling.md) in a future PR!

> Once NixOS is supported, we should remove the `gnome-text-editor` module completely and just use `lib.mkRenamedOptionModule` to handle its enable option.
> 
> Do you want to work on NixOS now, or merge this and leave it for later?

Now that you mention it's possible, I'd rather work on the NixOS implementation first, so there should be a seamless transition from the `gnome-text-editor` module.

Not sure if we should actually _remove_ the old module though, because some might have disabled it when its overlay caused issues. If we just remove the module, `targets.gnome-text-editor.enable = false` would result in an error, which might be confusing to some.

brckd commented:
@danth I added a NixOS implementation by overriding various `gtksourceview` packages. But now all other packages depending on `gtksourceview` will be rebuilt. I think this is quite a high cost for something that's already available as a Home Manager module.

The Stylix manual says:
> Many applications cannot be configured system wide, so Stylix will also need [Home Manager](https://github.com/nix-community/home-manager) to be able to change their settings within your home directory.
For this reason I would remove the NixOS module.

brckd commented:
For some reason, the theme is no longer present in the GNOME Text Editor.

brckd commented:
OK, apparently [this](https://gnome.pages.gitlab.gnome.org/gtksourceview/gtksourceview5/style-reference.html) is the updated documentation for gtksourceview-5:
> `${PREFIX}/share/gtksourceview-5/` (where ${PREFIX} can be `/usr/` or `/usr/local/` if you have installed from source)

I copied the theme to this location and `gtksourceview-2.0`. I also added it to `gtksourceview-4.0`, all though I couldn't find any documentation for this version.

The theme works again!

brckd commented:
Well, one problem with this approach is that this doesn't work if Steam hasn't downloaded its config files before.

trueNAHO commented:
> @danth I added a NixOS implementation by overriding various `gtksourceview` packages. But now all other packages depending on `gtksourceview` will be rebuilt. I think this is quite a high cost for something that's already available as a Home Manager module.
>
> The Stylix manual also says:
>
> > Many applications cannot be configured system wide, so Stylix will also need [Home Manager](https://github.com/nix-community/home-manager) to be able to change their settings within your home directory.
>
> Realistically, most Stylix users also use Home Manager anyways. For this reason I'll disable the NixOS module by default.

Is it not possible to enable the module when the Stylix Home Manager module is not used? For reference, commit 2c20aed3b39a ("qt: autoenable hm module only on NixOS (#942)") uses `osConfig` as a workaround for detecting this.

brckd commented:
Sure, that would be possible! However this wouldn't be able to detect standalone Home Manager.

awwpotato commented on modules/gtksourceview/nixos.nix:
this should be using the new [overlays system](https://stylix.danth.me/modules.html#overlays)

brckd replied:
Oh right, fixed it! There should be some way to disable to overlay when using Home Manager though, because the Home Manager module does the same as the overlay, but doesn't need a rebuild. And because of #954.

awwpotato replied:
you could probably do something like `mkEnableTarget "" (config ? home)`

danth replied:
If there is a way to check the [module class](https://ryantm.github.io/nixpkgs/module-system/module-system/#module-system-lib-evalModules-param-class), that would be more robust

brckd replied:
I think I implemented a more elegant solution in [5abcc42](https://github.com/danth/stylix/pull/958/commits/5abcc42a4b172680302368083adf1bf96caf8e5f). I leave the enable option as is and only override the default within the home manager module.

trueNAHO replied:
> I think I implemented a more elegant solution in [9420557](https://github.com/danth/stylix/pull/958/commits/94205576010863aedbb3bcd9672843ab9eaea712). I leave the enable option as is and only override the default within the home manager module.

Should we mark this as resolved?

awwpotato commented on modules/gtksourceview/overlay.nix:
```diff
@@ -0,0 +1,38 @@
+  };
+in
+{
+  options.stylix.targets.gtksourceview = {
+    enable = config.lib.stylix.mkEnableTarget "GTKSourceView" (config ? home);
+    overlay.enable = mkEnableOption "the overlay";
+  };
```
shouldn't it be this?
```suggestion
  options.stylix.targets.gtksourceview = with config.lib.stylix; {
    enable = mkEnableTarget "GTKSourceView" true;
    overlay.enable = mkEnableTarget "the overlay" (! config ? home);
  };
```

brckd replied:
Yea right, allthough I'm not sure whether mkEnableTarget should be used for this, because the overlay isn't strictly a target.

awwpotato replied:
mkEnableTarget is just a wrapper over mkEnableOption:
```nix
humanName: autoEnable:
lib.mkEnableOption "theming for ${humanName}"
// {
  default = cfg.enable && cfg.autoEnable && autoEnable;
  example = !autoEnable;
}
// lib.optionalAttrs autoEnable {
  defaultText = lib.literalMD "same as `stylix.autoEnable`";
};
```

awwpotato replied:
hmmm actually your right it shouldn't be used

awwpotato replied:
it should be:
```
overlay.enable = mkEnableTarget "the overlay" // { default = !(config ? home); };

trueNAHO commented on modules/gtksourceview/meta.nix:
```diff
@@ -0,0 +1,4 @@
+{
+  maintainers = [ ];
```
Since commit e43eb4e2a7dfbd96454df2b1c9418299b4373773, new modules are required to have at least one maintainer. See commit 8f0ee5532506e0dc62922eefb8c83fa758bcc684 for examples on how to add yourself as a maintainer.

brckd replied:
Fixed, but see https://github.com/nix-community/stylix/pull/958#issuecomment-3038905789.

trueNAHO commented on modules/gnome-text-editor/hm.nix:
```diff
@@ -1,9 +1,32 @@
+                osConfig
+              ])
+            )
+            "stylix: gnome-text-editor: This module will probably not work because the `gtksourceview' target is not enabled.";
```
Nitpick:

```suggestion
            "stylix: gnome-text-editor: this module will probably not work because the `gtksourceview' target is not enabled";
```

brckd replied:
Removed this warning for now, see https://github.com/nix-community/stylix/pull/958#issuecomment-3038895191.

trueNAHO reviewed (changes requested):
@brckd, could you resolve the merge conflicts?

trueNAHO commented:
What is the status of this PR? At this point, I assume there are quiet many merge conflicts.

brckd commented:
I honestly can't remember because it's been so long. I'll try to tackle my backlog next week, but I can't promise anything. After that I will be working on a personal project for a year and I'm not sure how available I'll be during that time.

brckd commented:
I rebased this onto master and removed some checks to simplify things for now. I enabled the overlay by default because there aren't many apps that depend on gtksourceview, so the rebuild wouldn't be too expensive.

I removed the warning for gnome-text-editor that it depends on gtksourceview, because gtksourceview is enabled by default anyways. There should probably be a more streamlined way to declare dependencies.

brckd commented:
As per https://github.com/nix-community/stylix/pull/958#issuecomment-3037193959 I can only provide maintenance on a best-efford basis. These modules should be fine though, since they aren't too complex.

brckd commented:
The maintainers file should comment that it's merged with `nixpkgs` maintainers.

brckd commented:
Seems to work!
![image](https://github.com/user-attachments/assets/724b2382-6e51-48b1-a52f-36d3154bc4a6)

awwpotato commented on modules/gnome-text-editor/meta.nix:
```diff
@@ -1,5 +1,7 @@
+{ lib, ... }:
+
```
```suggestion
```

awwpotato commented on modules/gtksourceview/hm.nix:
can we please use mkTarget here

brckd replied:
Oops, I'm still used to the old way of doing things™ :sweat_smile: Fixed it though!

awwpotato commented on modules/gtksourceview/overlay.nix:
```diff
@@ -0,0 +1,38 @@
+  };
+in
+{
+  options.stylix.targets.gtksourceview = {
+    enable = mkEnableTarget "GTKSourceView" true;
+  };
+
+  overlay =
+    _final: prev:
+    optionalAttrs (config.stylix.enable && cfg.enable) {
```
would ideal come from mkTarget
```suggestion
  overlay =
    _final: prev:
    optionalAttrs (config.stylix.enable && options.stylix.targets ? enable && cfg.enable) {
```

awwpotato replied:
@brckd this needs to be added, the current code will cause an error

brckd replied:
I hope that [3bfdcc1](https://github.com/nix-community/stylix/pull/958/commits/3bfdcc1c0b0ae92a10737b5bb010e64fff0f91b2) fixes it. Not sure how to test that though.

awwpotato reviewed (changes requested):
could you also add a testbed for both hm and nixos

brckd commented:
GTKSourceView isn't an app, but multiple apps depend on it. There is already a testbed for the GNOME Text Editor.

awwpotato reviewed (approved):
> GTKSourceView isn't an app, but multiple apps depend on it. There is already a testbed for the GNOME Text Editor.

Oh, I see. In that case diff LGTM

trueNAHO commented on modules/gtksourceview/hm.nix:
```diff
@@ -0,0 +1,23 @@
+{ mkTarget, ... }:
+
```
```suggestion
```

trueNAHO commented on modules/gtksourceview/hm.nix:
```diff
@@ -0,0 +1,23 @@
+{ mkTarget, ... }:
+
+let
+  style = {
+    template = ./template.xml.mustache;
+    extension = ".xml";
+  };
+in
+mkTarget {
+  name = "gtksourceview";
+  humanName = "GTKSourceView";
+
+  configElements =
+    { colors, ... }:
+    {
+      xdg.dataFile = {
+        "gtksourceview-2.0/styles/stylix.xml".source = colors style;
+        "gtksourceview-3.0/styles/stylix.xml".source = colors style;
+        "gtksourceview-4/styles/stylix.xml".source = colors style;
+        "gtksourceview-5/styles/stylix.xml".source = colors style;
+      };
```
What about generating the files from a list of supported versions:

```suggestion
mkTarget {
  name = "gtksourceview";
  humanName = "GTKSourceView";

  configElements =
    { colors, ... }:
    {
      xdg.dataFile = builtins.listToAttrs (
        map
          (
            version:
            lib.nameValuePair "gtksourceview-${version}/styles/stylix.xml" {
              source = colors {
                template = ./template.xml.mustache;
                extension = ".xml";
              };

            }
          )
          [
            "2.0"
            "3.0"
            "4"
            "5"
          ]
      );
```


This makes it easier to add or remove versions in the future. The same pattern is already used in the Blender module:

https://github.com/nix-community/stylix/blob/82a1f36f80a0bdc9875c8b868ea6fe930b09887c/modules/blender/hm.nix#L28-L47

Do not forget to add `lib` into the namespace:

```diff
-{ mkTarget, ... }:
+{ mkTarget, lib, ... }:
```

brckd replied:
I personally think that the added complexity and decreased readability is not worth it here. We can change it in the future in case we actually need to add new versions.

trueNAHO commented on modules/gtksourceview/overlay.nix:
```diff
@@ -0,0 +1,38 @@
+{ config, lib, ... }:
+
```
```suggestion
```

trueNAHO commented on modules/gtksourceview/hm.nix:
```diff
@@ -0,0 +1,29 @@
+                template = ./template.xml.mustache;
+                extension = ".xml";
+              };
+
```
```suggestion
```

trueNAHO reviewed (approved):
LGTM.
